File: 871-minimum-number-of-refueling-stops/871-minimum-number-of-refueling-stops.py
class Solution:
    def minRefuelStops(self, target: int, startFuel: int, stations: List[List[int]]) -> int:
        # dp[i][j] means the max dis we can reach with j stops in ith stations
        n = len(stations)
        
        dp = [[0] * (n + 1) for _ in range(n + 1)]
        for i in range(n + 1):
            dp[i][0] = startFuel
            
        for i in range(1, n + 1):
            for j in range(1, i + 1):
                # don't stop at the curr station, the dis 
                dp[i][j] = dp[i - 1][j]
                # we can reach the station
                if (dp[i - 1][j - 1] >= stations[i - 1][0]):
                    dp[i][j] = max(dp[i][j], dp[i - 1][j - 1] + stations[i - 1][1])
        
        for i, d in enumerate(dp[n]):
            if d >= target:
                return i
        return -1
        

File: 871-minimum-number-of-refueling-stops/871-minimum-number-of-refueling-stops.cpp
typedef unsigned long long ull;
class Solution {
public:
    int minRefuelStops(int target, int startFuel, vector<vector<int>>& stations) {
        int n = stations.size();
        // dp[i][j] use up to ith out of j stations, the farest dis we can reach
        vector<vector<ull>> dp(n + 1, vector<ull>(n + 1, 0));
        
        // stop 0 station, the farest location is the init fuel
        for (int i = 0; i < n + 1; ++i) dp[i][0] = startFuel;
        
        for (int i = 1; i < n + 1; ++i) {
            for (int j = 1; j < i + 1; ++j) {
                dp[i][j] = dp[i - 1][j];
                
                if (dp[i - 1][j - 1] >= stations[i - 1][0]) {
                    dp[i][j] = max(dp[i][j], dp[i - 1][j - 1] + stations[i - 1][1]);
                }
            }
        }
        for (int i = 0; i < n + 1; ++i) {
            if (dp[n][i] >= target) return i;
        }
        return -1;
    }
};